File: kernels/src/kernels/cli/skills.py
import shutil
import sys
from argparse import Namespace
from pathlib import Path

from huggingface_hub.utils import get_session

DEFAULT_SKILL_ID = "cuda-kernels"
SUPPORTED_SKILL_IDS = ("cuda-kernels", "rocm-kernels")
_GITHUB_RAW_BASE_TEMPLATE = (
    "https://raw.githubusercontent.com/huggingface/kernels/main/skills/{skill_id}"
)
_LOCAL_SKILLS_DIR = Path(__file__).resolve().parents[4] / "skills"

GLOBAL_TARGETS = {
    "codex": Path("~/.codex/skills"),
    "claude": Path("~/.claude/skills"),
    "opencode": Path("~/.config/opencode/skills"),
}

LOCAL_TARGETS = {
    "codex": Path(".codex/skills"),
    "claude": Path(".claude/skills"),
    "opencode": Path(".opencode/skills"),
}


def _download(url: str) -> str:
    response = get_session().get(url)
    response.raise_for_status()
    return response.text


def _github_raw_base(skill_id: str) -> str:
    return _GITHUB_RAW_BASE_TEMPLATE.format(skill_id=skill_id)


def _manifest_url(skill_id: str) -> str:
    return f"{_github_raw_base(skill_id)}/manifest.txt"


def _local_skill_root(skill_id: str) -> Path:
    return _LOCAL_SKILLS_DIR / skill_id


def _download_manifest(skill_id: str) -> list[str]:
    entries: list[str] = []
    try:
        raw_manifest = _download(_manifest_url(skill_id))
    except Exception:
        local_manifest = _local_skill_root(skill_id) / "manifest.txt"
        if not local_manifest.exists():
            raise
        raw_manifest = local_manifest.read_text(encoding="utf-8")

    for line in raw_manifest.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        entries.append(stripped)
    return entries


def _download_file(skill_id: str, rel_path: str) -> str:
    try:
        return _download(f"{_github_raw_base(skill_id)}/{rel_path}")
    except Exception:
        local_file = _local_skill_root(skill_id) / rel_path
        if local_file.exists():
            return local_file.read_text(encoding="utf-8")
        raise


def _remove_existing(path: Path) -> None:
    if path.is_symlink() or path.is_file():
        path.unlink()
        return

    shutil.rmtree(path)


def _install_to(target: Path, force: bool, skill_id: str) -> Path:
    target = target.expanduser().resolve()
    target.mkdir(parents=True, exist_ok=True)
    dest = target / skill_id

    if dest.exists():
        if not force:
            raise SystemExit(
                f"Skill already exists at {dest}.\n" "Re-run with --force to overwrite."
            )
        _remove_existing(dest)

    for rel_path in _download_manifest(skill_id):
        content = _download_file(skill_id, rel_path)
        output_file = dest / rel_path
        output_file.parent.mkdir(parents=True, exist_ok=True)
        output_file.write_text(content, encoding="utf-8")

    return dest


def add_skill(args: Namespace) -> None:
    skill_id = getattr(args, "skill", DEFAULT_SKILL_ID)
    if skill_id not in SUPPORTED_SKILL_IDS:
        supported = ", ".join(SUPPORTED_SKILL_IDS)
        raise SystemExit(f"Unsupported skill '{skill_id}'. Supported skills: {supported}")

    if not (args.claude or args.codex or args.opencode or args.dest):
        print(
            "Pick a destination via --claude, --codex, --opencode, or --dest.",
            file=sys.stderr,
        )
        sys.exit(1)

    targets_dict = GLOBAL_TARGETS if args.global_ else LOCAL_TARGETS
    targets: list[Path] = []

    if args.claude:
        targets.append(targets_dict["claude"])
    if args.codex:
        targets.append(targets_dict["codex"])
    if args.opencode:
        targets.append(targets_dict["opencode"])
    if args.dest:
        targets.append(args.dest)

    for target in targets:
        installed_path = _install_to(target, force=args.force, skill_id=skill_id)
        print(f"Installed '{skill_id}' to {installed_path}")

File: kernel-builder/src/skills.rs
use std::fs;
use std::path::PathBuf;

use eyre::{Context, Result};

pub const DEFAULT_SKILL_ID: &str = "cuda-kernels";
pub const SUPPORTED_SKILL_IDS: &[&str] = &["cuda-kernels", "rocm-kernels"];

const GITHUB_RAW_BASE_TEMPLATE: &str =
    "https://raw.githubusercontent.com/huggingface/kernels/main/kernel-builder/skills";

struct Targets {
    codex: PathBuf,
    claude: PathBuf,
    opencode: PathBuf,
}

fn global_targets() -> Result<Targets> {
    let home = dirs::home_dir().ok_or_else(|| eyre::eyre!("Cannot determine home directory"))?;
    Ok(Targets {
        codex: home.join(".codex/skills"),
        claude: home.join(".claude/skills"),
        opencode: home.join(".config/opencode/skills"),
    })
}

fn local_targets() -> Targets {
    Targets {
        codex: PathBuf::from(".codex/skills"),
        claude: PathBuf::from(".claude/skills"),
        opencode: PathBuf::from(".opencode/skills"),
    }
}

fn github_raw_base(skill_id: &str) -> String {
    format!("{GITHUB_RAW_BASE_TEMPLATE}/{skill_id}")
}

fn download(url: &str) -> Result<String> {
    let body = ureq::get(url)
        .call()
        .wrap_err_with(|| format!("Failed to fetch {url}"))?
        .into_body()
        .read_to_string()
        .wrap_err("Failed to read response body")?;
    Ok(body)
}

fn download_manifest(skill_id: &str) -> Result<Vec<String>> {
    let url = format!("{}/manifest.txt", github_raw_base(skill_id));
    let raw = download(&url)?;
    let entries: Vec<String> = raw
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .map(String::from)
        .collect();
    Ok(entries)
}

fn download_file(skill_id: &str, rel_path: &str) -> Result<String> {
    let url = format!("{}/{rel_path}", github_raw_base(skill_id));
    download(&url)
}

fn remove_existing(path: &PathBuf) -> Result<()> {
    if path.is_symlink() || path.is_file() {
        fs::remove_file(path)
            .wrap_err_with(|| format!("Cannot remove {}", path.display()))?;
    } else if path.is_dir() {
        fs::remove_dir_all(path)
            .wrap_err_with(|| format!("Cannot remove directory {}", path.display()))?;
    }
    Ok(())
}

fn install_to(target: &PathBuf, force: bool, skill_id: &str) -> Result<PathBuf> {
    let target = fs::canonicalize(target).unwrap_or_else(|_| target.clone());

    fs::create_dir_all(&target)
        .wrap_err_with(|| format!("Cannot create directory {}", target.display()))?;

    let dest = target.join(skill_id);

    if dest.exists() {
        if !force {
            eyre::bail!(
                "Skill already exists at {}.\nRe-run with --force to overwrite.",
                dest.display()
            );
        }
        remove_existing(&dest)?;
    }

    let manifest = download_manifest(skill_id)?;
    for rel_path in &manifest {
        let content = download_file(skill_id, rel_path)?;
        let output_file = dest.join(rel_path);
        if let Some(parent) = output_file.parent() {
            fs::create_dir_all(parent)
                .wrap_err_with(|| format!("Cannot create directory {}", parent.display()))?;
        }
        fs::write(&output_file, &content)
            .wrap_err_with(|| format!("Cannot write {}", output_file.display()))?;
    }

    Ok(dest)
}

pub fn add_skill(
    skill_id: &str,
    claude: bool,
    codex: bool,
    opencode: bool,
    global: bool,
    dest: Option<PathBuf>,
    force: bool,
) -> Result<()> {
    if !SUPPORTED_SKILL_IDS.contains(&skill_id) {
        let supported = SUPPORTED_SKILL_IDS.join(", ");
        eyre::bail!("Unsupported skill '{skill_id}'. Supported skills: {supported}");
    }

    if !claude && !codex && !opencode && dest.is_none() {
        eyre::bail!("Pick a destination via --claude, --codex, --opencode, or --dest.");
    }

    let mut install_targets: Vec<PathBuf> = Vec::new();

    if global {
        let targets = global_targets()?;
        if claude {
            install_targets.push(targets.claude);
        }
        if codex {
            install_targets.push(targets.codex);
        }
        if opencode {
            install_targets.push(targets.opencode);
        }
    } else {
        let targets = local_targets();
        if claude {
            install_targets.push(targets.claude);
        }
        if codex {
            install_targets.push(targets.codex);
        }
        if opencode {
            install_targets.push(targets.opencode);
        }
    }

    if let Some(d) = dest {
        install_targets.push(d);
    }

    for target in &install_targets {
        let installed_path = install_to(target, force, skill_id)?;
        println!(
            "Installed '{}' to {}",
            skill_id,
            installed_path.display()
        );
    }

    Ok(())
}
